syntax = "proto3";

service DistributedBookstore {
  rpc CreateChain(CreateChainRequest) returns (CreateChainResponse) {}
  rpc Link(LinkRequest) returns (LinkResponse) {}
  rpc ListChain(ListChainRequest) returns (ListChainResponse) {}

  rpc ListBooks(ListBooksRequest) returns (ListBooksResponse) {}
  rpc Read(ReadRequest) returns (ReadResponse) {}
  rpc Write(WriteRequest) returns (WriteResponse) {}
  rpc Clean(CleanRequest) returns (CleanResponse) {}
}

message CreateChainRequest {}

message CreateChainResponse {
  repeated string process_list = 1;
}

message LinkRequest {
  string head = 1;
  string tail = 2;
  repeated ChainNode chain_nodes = 3;
}

message ChainNode {
  string process_id = 1;
  string successor_id = 2;
  string predecessor_id = 3;
}

message LinkResponse {}

message ListChainRequest {}

message ListChainResponse {
  repeated ChainNode chain_nodes = 1;
}

message Book {
  string name = 1;
  double price = 2;
}

message ListBooksRequest {}

message ListBooksResponse {
  repeated Book books = 1;
}

message ReadRequest {
  string name = 1;
}

message ReadResponse {
  Book book = 1;
}

message WriteRequest {
  string process_id = 1;
  string name = 2;
  double price = 3;
}

message WriteResponse {}

message CleanRequest {
  string process_id = 1;
  Book book = 2;
}

message CleanResponse {}
